fix stuck reconnection problem

dharjeezy opened the pull request:
When the `reconnecting_rpc_client` detects a WebSocket disconnection and enters its reconnection logic, existing RPC subscriptions are not terminated. The background tasks (subscription_handler) for these subscriptions remain stucked, awaiting messages from a dead connection.

This PR introduces a mechanism to gracefully terminate active subscriptions when a client reconnect begins.

dharjeezy committed:
fix stuck reconnection problem

dharjeezy committed:
test

dharjeezy committed:
make use of a channel in the test to signal termination
fix initial failing test

dharjeezy committed:
use jsonrpsee server as dev dependency

dharjeezy commented on rpcs/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ description = "Make RPC calls to Substrate based nodes"
 keywords = ["parity", "subxt", "rpcs"]
 
 [features]
-default = ["jsonrpsee", "native"]
```
I have reverted on this

jsdw commented:
Hey @dharjeezy, had a more proper look and the logic all sounds sane to me! Previously subscriptions would just return `DisconnectedWillReconnect` to users, and I agree that this feels wrong since they are not actually reconnected, so returning `None` there makes sense, as does using the Notify stuff to tell subscriptions to end in this case. Just a tiny nit to close a perhaps very unlikely race :)

dharjeezy committed:
create listener future once
clippy

dharjeezy commented:
> Hey @dharjeezy, had a more proper look and the logic all sounds sane to me! Previously subscriptions would just return `DisconnectedWillReconnect` to users, and I agree that this feels wrong since they are not actually reconnected, so returning `None` there makes sense, as does using the Notify stuff to tell subscriptions to end in this case. Just a tiny nit to close a perhaps very unlikely race, and by the looks of it clippy is moaning :)

Thank you for your prompt response, I have now addressed your concerns.

jsdw commented:
Argh, new version of Rust released today which means new clippy errors; let me fix those on master quickly and you can merge that in :)

jsdw commented:
@dharjeezy master is ready to be merged into here now to hopefully fix clippy errors!

dharjeezy committed:
Merge branch 'paritytech:master' into dami/fix-reconnection-problem

dharjeezy commented:
@jsdw PR is now in sync with master

dharjeezy committed:
fix clippy

dharjeezy committed:
patience in test

dharjeezy commented:
@jsdw for some weird reason, 2 tests in CI is failing, but when i run these tests locally it passes. I tried introducing some delay before assertion. Can you help retrigger CI?

jsdw commented:
CI is often wayy slower than local so I'm not too surprised :)

dharjeezy committed:
longer wait time

dharjeezy committed:
deterministic Event Channel for testing purpose only

dharjeezy committed:
fmt

dharjeezy committed:
revert

dharjeezy committed:
yield now

jsdw commented:
Closing in favour of #2057
